srclib 0.0.37 (new formula)

samertm opened the pull request:
Add srclib to Homebrew. srclib is a language-agnostic language analysis tool. srclib installs as the binary 'src', which conflicts with the 'src' formula. Is there an issue with that? Or is there some way 'src' and 'srclib' can coexist?

I went through some pull requests for other Go projects, and it seems like the idiomatic thing is to add all of your dependencies as `go_resource` statements rather than rely on the `go` tool pulling down dependencies during the install step, so I wrote a tool that generates the statements: https://github.com/samertm/homebrew-go-resources Do you think I should mention it Formula Cookbook README?

I know testing "\<binary\> version" is frowned upon, but I couldn't figure out what else to test `src` on, because it relies on entire repositories as input, and you need to download a toolchain before you can do anything interesting with it. Thoughts on that?

Let me know if I did anything wrong! Thanks in advance for looking at this.

Thanks,
Samer

mikemcquaid commented on Library/Formula/srclib.rb:
```diff
@@ -0,0 +1,150 @@
+  end
+
+  test do
+    result = pipe_output("#{bin}/src version")
```
Can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

mikemcquaid commented:
> I know testing "<binary> version" is frowned upon, but I couldn't figure out what else to test src on, because it relies on entire repositories as input, and you need to download a toolchain before you can do anything interesting with it. Thoughts on that?

Can you elaborate a bit on that? Could we create e.g. a git repository locally? If it needs to have a toolchain downloaded should we do that for the users?

DomT4 commented on Library/Formula/srclib.rb:
```diff
@@ -0,0 +1,150 @@
+  depends_on "go" => :build
+
+  go_resource "code.google.com/p/rog-go" do
+    url "https://code.google.com/p/rog-go", :revision => "7088342b70fc1995ada4986ef2d093f340439c78", :using => :hg
```
Might be nice to bump this onto multiple lines for the sake of readability, something like:

```ruby
go_resource "code.google.com/p/rog-go" do
  url "https://code.google.com/p/rog-go",
       :revision => "7088342b70fc1995ada4986ef2d093f340439c78", :using => :hg
end
```

Or something similar may work.

samertm replied:
Done, thanks.

samertm commented:
srclib is a generic interface for working with language analysis, but it needs language-specific toolchains in order to actual analyze code. So, if you wanted to use srclib to analyze a Go project, you would install the toolchain first with "src toolchain install go", and then you could run normal srclib commands. Toolchains exist for Go, Python, Ruby, Haskell, Java, and JavaScript, and I don't think it makes sense to install a set of them automatically because it's valid to only want one or two toolchains. srclib has an "install-std" command that pulls down Go, Python, Ruby, and JavaScript if the user wants a more complete set, and it's easy enough to install toolchains that you want.

I'm not sure how complete the tests need to be (we aren't just testing existence, but semi-functioning? Does the entire test suite need to run?), but there's also a "sample" toolchain which tests that src works at all and always outputs the same thing. I can add that toolchain and create a sample repo, take a look at the output and see that it matches something basic, and then delete the sample toolchain.

mistydemeo commented:
> srclib installs as the binary 'src', which conflicts with the 'src' formula. Is there an issue with that? Or is there some way 'src' and 'srclib' can coexist?

There are two options here:

* Mark the two formulae as `conflicts_with` each other, which prevents them from being linked side-by-side
* Rename the binary in this formula

samertm commented:
@mistydemeo thanks, added the `conflicts_with` statements.

mikemcquaid commented:
> I'm not sure how complete the tests need to be (we aren't just testing existence, but semi-functioning? Does the entire test suite need to run?), but there's also a "sample" toolchain which tests that src works at all and always outputs the same thing. I can add that toolchain and create a sample repo, take a look at the output and see that it matches something basic, and then delete the sample toolchain

Yeh, that sounds like a good idea. Assuming it's a `src toolchain install sample` call: you can have the actual call in the test. Also, the test is done in a temporary directory and overrides `$HOME` too and then deletes both after completion.

mistydemeo commented on Library/Formula/src.rb:
```diff
@@ -11,6 +11,8 @@ class Src < Formula
     sha1 "cd51d4f5714a360549d9548a25c27a508aa9ae1b" => :mountain_lion
   end
 
+  conflicts_with "src", :because => "both install a 'src' binary"
```
Should be `srclib`?

samertm replied:
Fixed, thanks.

tdsmith commented:
>  I wrote a tool that generates the statements: https://github.com/samertm/homebrew-go-resources Do you think I should mention it Formula Cookbook README?

Nice, thanks. I added a plug in 8f69a54; feel free to send a PR if you think I'm selling it short. :)

samertm commented:
@tdsmith thanks :) that looks good to me.
@mistydemeo :x my bad, that was a hasty mistake.
@mikemcquaid that sounds good to me, will get to it tonight.

samertm commented:
@mikemcquaid "src" needs to be in PATH to work correctly, because it generates makefiles that rely on "src" being available unqualified. Is that okay? I can update it to use the fully qualified path if that's an issue. https://github.com/Homebrew/homebrew/pull/37651/files#diff-c7ff92fc96b077a7564075c03a2f5af8R181

mikemcquaid commented:
@samertm Yeh, that's fine, just add it to the PATH explicitly in the test in that case (`ENV["PATH"]`)

samertm commented:
@mikemcquaid hey, sorry for taking ages with this! The test was failing on Mavericks because srclib was mistakenly touching "$HOME/.gitignore", but I've fixed it & updated all of the dependencies. I've changed the version to srclib 0.0.41, should I open a new pull request?

mistydemeo commented on Library/Formula/srclib.rb:
```diff
@@ -0,0 +1,192 @@
+  end
+
+  test do
+    system "git", "clone", "https://github.com/sourcegraph/srclib-sample"
```
Make this a regular `resource` and stage it in the test block rather than directly doing `git clone`, please!

samertm replied:
Fixed (I believe), thanks!

samertm commented:
@mistydemeo @mikemcquaid I made that change and squashed the branch to a single commit. Looks good for merge? Thanks again for the help!

samertm commented:
Ping @mistydemeo @mikemcquaid. Looks good to merge? :D

mikemcquaid commented:
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

samertm commented:
Thanks for the help :D
